src/datetime.rs

```diff
@@ -926,77 +948,61 @@ pub mod serde {
         /// # #[macro_use] extern crate serde_derive;
         /// # #[macro_use] extern crate serde_json;
         /// # extern crate chrono;
-        /// # use chrono::{TimeZone, DateTime, Utc};
-        /// use chrono::serde::ts_seconds::serialize as to_ts;
-        /// #[derive(Serialize)]
+        /// # use chrono::{DateTime, Utc};
+        /// use chrono::serde::ts_nanoseconds::deserialize as from_nano_ts;
+        /// #[derive(Deserialize)]
         /// struct S {
-        ///     #[serde(serialize_with = "to_ts")]
+        ///     #[serde(deserialize_with = "from_nano_ts")]
         ///     time: DateTime<Utc>
         /// }
         ///
-        /// # fn example() -> Result<String, serde_json::Error> {
-        /// let my_s = S {
-        ///     time: Utc.ymd(2015, 5, 15).and_hms(10, 0, 0),
-        /// };
-        /// let as_string = serde_json::to_string(&my_s)?;
-        /// assert_eq!(as_string, r#"{"time":1431684000}"#);
-        /// # Ok(as_string)
+        /// # fn example() -> Result<S, serde_json::Error> {
+        /// let my_s: S = serde_json::from_str(r#"{ "time": 1526522699918355733 }"#)?;
+        /// # Ok(my_s)
         /// # }
         /// # fn main() { example().unwrap(); }
         /// ```
-        pub fn serialize<S>(dt: &DateTime<Utc>, serializer: S) -> Result<S::Ok, S::Error>
-            where S: ser::Serializer
+        pub fn deserialize<'de, D>(d: D) -> Result<DateTime<Utc>, D::Error>
+            where D: de::Deserializer<'de>
         {
-            serializer.serialize_i64(dt.timestamp())
+            Ok(try!(d.deserialize_i64(NanoSecondsTimestampVisitor)))
         }
 
-        struct SecondsTimestampVisitor;
+        struct NanoSecondsTimestampVisitor;
 
-        impl<'de> de::Visitor<'de> for SecondsTimestampVisitor {
-            type Value = DateTime<FixedOffset>;
+        impl<'de> de::Visitor<'de> for NanoSecondsTimestampVisitor {
+            type Value = DateTime<Utc>;
 
             fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result
             {
                 write!(formatter, "a unix timestamp in seconds")
             }
 
             /// Deserialize a timestamp in seconds since the epoch
-            fn visit_i64<E>(self, value: i64) -> Result<DateTime<FixedOffset>, E>
+            fn visit_i64<E>(self, value: i64) -> Result<DateTime<Utc>, E>
                 where E: de::Error
             {
-                from(FixedOffset::east(0).timestamp_opt(value, 0), &value)
+                serde_from(Utc.timestamp_opt(value / 1_000_000_000,
+                                            (value % 1_000_000_000) as u32),
```

Comment:
This fails if value is negative, because the modulus will return a negative number and casting it to u32 will make it exceed the maximal value of 999_999_999.
The same applies for the other deserializations of i64.

Reply:
I mean, that's how it was before. I didn't actually change that code, Git just says I did.

Would it be better to call `abs` on `value` before the modulus?

Reply:
I know that it is a previous issue. I just though I mention it, if the whole serialization part is changed here anyways.

I do not think `abs` alone works right. Suppose value is `-1`, then you want the two values passed to `timestamp_opt` to be `-1` and `999_999_999`, but just `abs` would give you `0` and `1`.

Reply:
As the nanosecond argument of `timestamp_opt` is unsigned, there isn’t a way to directly pass a negative timestamp correctly anyway. 

What I could do is check for a negative timestamp and offset the seconds by negative one second, then pass one second subtract the positive nanoseconds part for the nanoseconds argument? That would give the desired results.

Reply:
Same with the other signed timestamps.

Reply:
Sorry about the delay reviewing this!

That does seem reasonable, not handling negative timestamps is a... sad mistake, that I think might be duplicated through more of the codebase. Since this isn't you're fault I'm happy to merge this and fix the negative timestamp issue in a separate PR. Let me know what you think.

Reply:
Relatedly, I started trying to deal with negative timestamps in #252 but haven't had the time to finish it.